mmengine/infer/infer.py

```diff
@@ -0,0 +1,502 @@
+            device = get_device()
+
+        self.cfg = cfg
+        self.model = self._init_model(cfg, weights, device)  # type: ignore
+        self.pipeline = self._init_pipeline(cfg)
+        self.collate_fn = self._init_collate(cfg)
+        self.visualizer = self._init_visualizer(cfg)
+
+    def __call__(
+        self,
+        inputs: InputsType,
+        return_datasamples: bool = False,
+        batch_size: int = 1,
+        **kwargs,
+    ) -> dict:
+        """Call the inferencer.
+
+        Args:
+            inputs (InputsType): Inputs for the inferencer.
+            return_datasamples (bool): Whether to return results as
+                :obj:`BaseDataElement`. Defaults to False.
+            batch_size (int): Batch size. Defaults to 1.
+            **kwargs: Key words arguments passed to :meth:`preprocess`,
+                :meth:`forward`, :meth:`visualize` and :meth:`postprocess`.
+                Each key in kwargs should be in the corresponding set of
+                ``preprocess_kwargs``, ``forward_kwargs``, ``visualize_kwargs``
+                and ``postprocess_kwargs``.
+
+        Returns:
+            dict: Inference and visualization results.
+        """
+        (
+            preprocess_kwargs,
+            forward_kwargs,
+            visualize_kwargs,
+            postprocess_kwargs,
+        ) = self._dispatch_kwargs(**kwargs)
+
+        inputs = self.preprocess(
+            inputs, batch_size=batch_size, **preprocess_kwargs)
+        preds = []
+        for data in track(inputs, description='Inference'):
+            preds.extend(self.forward(data, **forward_kwargs))
+        visualization = self.visualize(
+            inputs, preds, **visualize_kwargs)  # type: ignore  # noqa: E501
+        results = self.postprocess(preds, visualization, return_datasamples,
+                                   **postprocess_kwargs)
+        return results
+
+    def preprocess(self, inputs: InputsType, batch_size: int = 1, **kwargs):
+        """Process the inputs into a model-feedable format.
+
+        Customize your preprocess by overriding this method. Preprocess should
+        return an iterable object, of which each item will be used as the
+        input of ``model.test_step``.
+
+        ``BaseInferencer.preprocess`` will return an iterable chunked data,
+        which will be used in __call__ like this:
+
+        .. code-block:: python
+
+            def __call__(self, inputs, batch_size=1, **kwargs):
+                chunked_data = self.preprocess(inputs, batch_size, **kwargs)
+                for batch in chunked_data:
+                    preds = self.forward(batch, **kwargs)
+
+        Args:
+            inputs (InputsType): Inputs given by user.
+            batch_size (int): batch size. Defaults to 1.
+
+        Yields:
+            Any: Data processed by the ``pipeline`` and ``collate_fn``.
+        """
+        if isinstance(inputs, str):
+            backend = get_file_backend(inputs)
```

Comment:
`get_file_backend` sometimes requires `backend_args` kwargs so should we make preprocess accept `backend_args`?